Qt: Add option to disable the system tray icon

Tyler-Hardin opened the pull request:
My changes leave all tray icon and menu creation/initialization logic untouched. It only shows or hides the icon according to the setting.

A new checkbox was added to the OptionsDialog under the Window tab. A bool option named "disableTray" was added to OptionsModel. This checkbox was mapped like other all options to the OptionsModel.

A signal was added to the OptionsModel for broadcasting changes the the
disableTray option. This signal was connected to a new slot added to
BitcoinGUI named updateDisableTray(bool). The slot simply hides or shows
the trayIcon in BitcoinGUI according to the parameter recieved.

The negative wording of "disableTray" instead of a positive wording
("enableTray") because enabled is the default. Perhaps "hideTray" would
be better than disable.

This pull request intends to resolve this issue: https://github.com/bitcoin/bitcoin/issues/7160

Tyler-Hardin commented:
Btw, this is my first pull request. Don't hurt me.

jonasschnelli commented on src/qt/forms/optionsdialog.ui:
```diff
@@ -20,7 +20,7 @@
     <widget class="QTabWidget" name="tabWidget">
      <property name="currentIndex">
-      <number>0</number>
+      <number>3</number>
```
I think this was done by Qt Creator. I guess it should still be `0`?

Tyler-Hardin commented:
Yes, it was. I'll edit it back to zero.

jonasschnelli commented:
Nice work!
Concept ACK will test in detail soon.

laanwj commented:
Concept ACK.

> Perhaps "hideTray" would be better than disable.

I'd prefer "setTrayVisible(bool)" - this is closer to Qt's own usage. I agree on the positive versus negative wording in principle, but on the other hand the default could change later, so may as well make it general.

Tyler-Hardin commented:
@laanwj, since we're getting verbose anyway, would "setTrayIconVisible(bool)" be even better? Because, technically, I'm setting the visibility on the TrayIcon.

Btw, thanks for reminding me about the setVisible function. I had used an `if(flag) tray->hide(); else tray->show()` logic that I can simplify to a single line with setVisible.

laanwj commented:
> setTrayIconVisible(bool)

Sure, even clearer.

Tyler-Hardin commented:
@laanwj, fixed.

jonasschnelli commented:
This currently results in a startup crash on OSX because `trayIcon` is NULL on OSX (see `void BitcoinGUI::createTrayIcon(const NetworkStyle *networkStyle)`).

IMO this PR should remove the `#ifndef Q_OS_MAC` for the tray-icon creation (so it also gets created on OSX) and disabled it by default on OSX (over platformstyle class).

laanwj commented:
@jonasschnelli I think the reason it's not created on OSX is because the menu items are added to some other menu (which effectively acts as a tray icon). It was non-trivial to get rid of the #ifdeffing there, at least it seemed back in the day when I implmemented platformstyle.

jonasschnelli commented:
utACK (testes a bit but will test more in detail).
I can fix the OSX tray-icon support later (if even possible).

jonasschnelli commented:
Needs squashing.

Tyler-Hardin commented:
@jonasschnelli, squashed.

jonasschnelli commented:
Tested ACK a8abd45ec7ddb16d09f9abfd90b6eea722c55103

A little nit:
If you have your tray icon disabled, it shortly flickers up during startup of Bitcoin-Qt.

A simple change can fix this:
```diff
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -545,7 +545,7 @@ void BitcoinGUI::createTrayIcon(const NetworkStyle *networkStyle)
     QString toolTip = tr("%1 client").arg(tr(PACKAGE_NAME)) + " " + networkStyle->getTitleAddText();
     trayIcon->setToolTip(toolTip);
     trayIcon->setIcon(networkStyle->getTrayAndWindowIcon());
-    trayIcon->show();
+    trayIcon->hide();
 #endif
 
     notificator = new Notificator(QApplication::applicationName(), trayIcon, this);
```

Tyler-Hardin committed:
Qt: Add option to hide the system tray icon

My changes leave all tray icon and menu creation/initialization logic
untouched. It only shows or hides the icon according to the setting.

A new checkbox was added to the OptionsDialog under the Window tab. A
bool option named "hideTrayIcon" was added to OptionsModel. This
checkbox was mapped like other all options to the OptionsModel.

A signal was added to the OptionsModel for broadcasting changes the the
hideTrayIcon option. This signal was connected to a new slot added to
BitcoinGUI named setTrayIconVisible(bool). The slot simply hides or
shows the trayIcon in BitcoinGUI according to the parameter recieved.

Tyler-Hardin commented:
@jonasschnelli, good catch. I added that edit.

jonasschnelli commented:
Tested ACK 8b0e4970281de15472a93a529302f3de15163943

laanwj commented:
Tested ACK 8b0e497
